pulsar-broker/src/test/java/org/apache/pulsar/broker/transaction/pendingack/PendingAckPersistentTest.java

```diff
@@ -345,4 +350,103 @@ private void testDeleteTopicThenDeletePendingAckManagedLedger() throws Exception
+        @Cleanup
+        Producer<byte[]> producer = pulsarClient.newProducer()
+                .topic(topic)
+                .sendTimeout(0, TimeUnit.SECONDS)
+                .enableBatching(false)
+                .create();
+
+        for (int i = 0; i < 20; i++) {
+            producer.newMessage().send();
+        }
+        // init
+        Message<byte[]> message = consumer.receive(5, TimeUnit.SECONDS);
+        Transaction transaction = pulsarClient.newTransaction()
+                .withTransactionTimeout(5, TimeUnit.SECONDS)
+                .build()
+                .get();
+        consumer.acknowledgeAsync(message.getMessageId(), transaction);
+        transaction.commit().get();
+
+        PersistentTopic persistentTopic = (PersistentTopic) getPulsarServiceList().get(0)
+                .getBrokerService().getTopic(topic, false).get().get();
+
+        PersistentSubscription persistentSubscription = persistentTopic.getSubscription(subName);
+        Field field = PersistentSubscription.class.getDeclaredField("pendingAckHandle");
+        field.setAccessible(true);
+        PendingAckHandleImpl pendingAckHandle = (PendingAckHandleImpl) field.get(persistentSubscription);
+        Field field1 = PendingAckHandleImpl.class.getDeclaredField("pendingAckStoreFuture");
+        field1.setAccessible(true);
+        PendingAckStore pendingAckStore = ((CompletableFuture<PendingAckStore>) field1.get(pendingAckHandle)).get();
+
+        Field field3 = MLPendingAckStore.class.getDeclaredField("pendingAckLogIndex");
+        Field field4 = MLPendingAckStore.class.getDeclaredField("upperLimitOfLogAppendTimes");
+
+        field3.setAccessible(true);
+        field4.setAccessible(true);
+
+        ConcurrentSkipListMap<PositionImpl, PositionImpl> pendingAckLogIndex =
+                (ConcurrentSkipListMap<PositionImpl, PositionImpl>) field3.get(pendingAckStore);
+        long upperLimitOfLogAppendTimes = (long) field4.get(pendingAckStore);
+        Assert.assertEquals(pendingAckLogIndex.size(), 1);
+        Assert.assertEquals(upperLimitOfLogAppendTimes, 5);
+
+        Awaitility.await().untilAsserted(() ->
+                Assert.assertEquals(persistentSubscription.getCursor().getPersistentMarkDeletedPosition().getEntryId(),
+                        ((MessageIdImpl)message.getMessageId()).getEntryId()));
+        // 7 more acks. Will find that there are still only two records in the map.
+        Transaction transaction1 = pulsarClient.newTransaction()
+                .withTransactionTimeout(5, TimeUnit.SECONDS)
+                .build()
+                .get();
+        //remove previous index
+        Message<byte[]> message0 = consumer.receive(5, TimeUnit.SECONDS);
+        consumer.acknowledgeAsync(message0.getMessageId(), transaction1).get();
+        Assert.assertEquals(pendingAckLogIndex.size(), 0);
+        upperLimitOfLogAppendTimes = (long) field4.get(pendingAckStore);
+        Assert.assertEquals(upperLimitOfLogAppendTimes, 5);
```

Comment:
also think about cursor persistentMarkerDeletePostion changed

Reply:
The transaction here is not committed, persistentMarkDeletePosition will not move, so there is no problem here